Add SDL_{Read,Write}PixelValue() functions (was Fix a number of OOB/Alignment/Endianness issues with reading/writing pixels)

sulix opened the pull request:
As @0x1F9F1 pointed out for ``SDL_PremultiplySurfaceAlphaToARGB8888()`` [here](https://github.com/libsdl-org/SDL/commit/84808ea4bbd12ac212ff7a4ef82d426f6d771e99#commitcomment-58499396), it's not correct to read a pixel from a non-32bit surface by taking a pointer to the pixel, casting it to a ``Uint32*``, dereferencing it, and using ``SDL_GetRGBA()``.

This is for three reasons:
- It could read beyond the bounds of the surface data for the last few pixels.
- Reads may not be 32-bit aligned, which can fault (or at least have performance problems) on some architectures.
- On big-endian systems, the correct pixel may be loaded into the higher-order bits of the ``Uint32``, which would lead to the wrong value being returned (either an adjacent pixel or garbage).

This wasn't actually causing any serious problems for ``SDL_PremultiplySurfaceAlphaToARGB8888()``, as the input is always ARGB8888 anyway, but it's worth having it be correct if we can. There also were a few other places in SDL (and, more particularly, the test programs), which had similar issues. This PR fixes all of the ones I found, by copying (and improving the 24-bit case of) the implementation in ``SDL_CalculateShapeBitmap()``.

There's probably a pretty good case for having some (probably user-accessible) PSet/PGet functions which do this for you, as it's quite easy to get wrong, and this code is now duplicated several times. This hopefully should do for now, though.

sezero commented:
Has this one been ever reviewed?

slouken commented on src/video/SDL_shape.c:
```diff
@@ -93,7 +93,11 @@ SDL_CalculateShapeBitmap(SDL_WindowShapeMode mode,SDL_Surface *shape,Uint8* bitm
                 case(3):
-                    pixel_value = *(Uint32*)pixel & (~shape->format->Amask);
+#if SDL_BYTEORDER == SDL_BIG_ENDIAN
+                    pixel_value = (pixel[0] << 16) | (pixel[1] << 8) | (pixel[2] << 0);
```
pixel[0] is Uint8, so shifting left 16 may leave it 0.

Was this code tested?

sulix replied:
I did test this, albeit only on little-endian machines. I did try reversing the byteorder check and making sure the colours came out reversed, though, so it seemed like it was working both ways.

My suspicion is that, yes, ``pixel[0]`` is promoted to ``int``, so this should work on any system with ``sizeof(int)`` ≥ 3. I'm not opposed to adding additional casts to handle old 16-bit compilers if you want, though.

slouken replied:
Let's cast it explicitly so it's clear what's happening here.

slouken commented:
With the amount of duplication, I agree, adding a utility function to get the pixel from a surface seems useful.

sulix commented:
> With the amount of duplication, I agree, adding a utility function to get the pixel from a surface seems useful.

Cool. Hopefully I'll have some time to put that together over the weekend.

sulix commented:
Updated and rebased this: it now adds two new functions: SDL_ReadPixelValue() and SDL_WritePixelValue(), which take a pointer and an ``SDL_PixelFormat`` and read or write a pixel value in ``Uint32`` form (as expected/produced by SDL_GetRGBA()/SDL_MapRGBA() respectively).

Note that these functions don't get the value directly from a surface yet: they just accept a pointer. This made more sense if we want to avoid repeated locking/unlocking of surfaces when looping over pixels, though I do plan to add functions which do take a surface and x/y coordinates as well if this seems useful.

slouken commented on include/SDL_pixels.h:
```diff
@@ -627,7 +627,35 @@ extern DECLSPEC void SDLCALL SDL_GetRGBA(Uint32 pixel,
+ * \sa SDL_MapRGB
+ * \sa SDL_MapRGBA
+ */
+extern DECLSPEC void SDLCALL SDL_WritePixelValue(void *data, SDL_PixelFormat *format, Uint32 value);
```
Minor detail, can you add a blank line here?

sulix replied:
I haven't got the time to make any of the more significant changes here right at the moment, but I have at least added the blank line. :-)

slouken commented on include/SDL_pixels.h:
```diff
@@ -627,7 +627,35 @@ extern DECLSPEC void SDLCALL SDL_GetRGBA(Uint32 pixel,
+ * \sa SDL_GetRGB
+ * \sa SDL_GetRGBA
+ */
+extern DECLSPEC Uint32 SDLCALL SDL_ReadPixelValue(void *data, SDL_PixelFormat *format);
```
It might be better to make these macros, to avoid per-pixel function call overhead.

sulix replied:
I'm not quite sure how to make this into a macro without making it horrendously ugly, given it ideally wants to "return" a value. Unless we wanted to have it accept a pointer to a Uint32 in which to place the result?

icculus commented:
Can we make sure there's a note in the doxygen that this is _not_ a fast way to access pixel data, and that people should be using something that works on blocks of data at a time, or the rendering APIs, or the surface APIs, or OpenGL, or literally anything but this?

slouken commented:
We could also just leave it in the internal SDL_pixels_c.h so we only use it because we know it's the right tool for the job.

icculus commented:
> We could also just leave it in the internal SDL_pixels_c.h so we only use it because we know it's the right tool for the job.

I agree.

sulix commented:
> Can we make sure there's a note in the doxygen that this is _not_ a fast way to access pixel data, and that people should be using something that works on blocks of data at a time, or the rendering APIs, or the surface APIs, or OpenGL, or literally anything but this?

No worries. I'll add a note saying that it's (a) too slow to use for general-purpose rendering, and (b) only really useful for getting/using a value for SDL_GetRGBA()/SDL_MapRGBA().

> We could also just leave it in the internal SDL_pixels_c.h so we only use it because we know it's the right tool for the job.

Would that be a problem for the uses of this in tests? There are enough of those I think we'll want _something_ exposed at least to tests. (And I'd sort-of rather have this function available to people who are trying to get pixel data to pass through SDL_{Get,Map}RGBA(), as most of the obvious alternatives are wrong, even if they are fast…)

The other option is, given this is going to be horribly slow anyway, to go all the way and have a full SDL_SurfacePGet()/SDL_SurfacePSet()-style interface based on surfaces, which at least won't have users brandishing pointers around as much…

slouken commented:
Okay, I think Ryan and I are leaning toward making this function either a horrible macro so people can use it more efficiently or keeping it entirely internal. Let's mull it over some more and we can merge the appropriate changes here.

Thanks!

1bsyl commented:
In case it helps, I am using this code for writing generic pixel routines. Without a too big overhead (eg, no per pixel function call, nor switch format). 
All macros get optimized at compile time because of const variables in the switch.

```

#define PIXEL_READER_BPP_4(ptr, x)          ((Uint32*)(ptr))[x]
#define PIXEL_WRITER_BPP_4(ptr, x, val)     ((Uint32*)(ptr))[x] = val;

#define PIXEL_READER_BPP_3(ptr, x)          ( ((Uint8*)(ptr))[3 * x]                    \
                                            | (((Uint8*)(ptr))[3 * x + 1] << 8)         \
                                            | (((Uint8*)(ptr))[3 * x + 2] << 16))

#define PIXEL_WRITER_BPP_3(ptr, x, val)     ((Uint8*)(ptr))[3 * x]     = val;           \
                                            ((Uint8*)(ptr))[3 * x + 1] = (val >> 8);    \
                                            ((Uint8*)(ptr))[3 * x + 2] = (val >> 16);

#define PIXEL_READER_BPP_2(ptr, x)          ((Uint16*)(ptr))[x]
#define PIXEL_WRITER_BPP_2(ptr, x, val)     ((Uint16*)(ptr))[x] = val;

#define PIXEL_READER_BPP_1(ptr, x)          ((Uint8*)(ptr))[x]
#define PIXEL_WRITER_BPP_1(ptr, x, val)     ((Uint8*)(ptr))[x] = val;


#define PIXEL_READER(ptr, x)                                                    \
   ((reader_bpp == 4) ? PIXEL_READER_BPP_4(ptr, x) :                            \
      ((reader_bpp == 3) ? PIXEL_READER_BPP_3(ptr, x) :                         \
         ((reader_bpp == 2) ? PIXEL_READER_BPP_2(ptr, x) :                      \
            ((reader_bpp == 1) ? PIXEL_READER_BPP_1(ptr, x) : (Uint32)0 ))))



#define PIXEL_WRITER(ptr, x, val)                       \
   if (writer_bpp == 4) {                               \
      PIXEL_WRITER_BPP_4(ptr, x, val)                   \
   }                                                    \
   if (writer_bpp == 3) {                               \
      PIXEL_WRITER_BPP_3(ptr, x, val)                   \
   }                                                    \
   if (writer_bpp == 2) {                               \
      PIXEL_WRITER_BPP_2(ptr, x, val)                   \
   }                                                    \
   if (writer_bpp == 1) {                               \
      PIXEL_WRITER_BPP_1(ptr, x, val)                   \
   }



#define PIXEL_SWITCH_BPP(code)              \
      switch (bpp) {                        \
         case 4:                            \
            {                               \
               const int reader_bpp = 4;    \
               const int writer_bpp = 4;    \
               code                         \
            }                               \
            break;                          \
         case 3:                            \
            {                               \
               const int reader_bpp = 3;    \
               const int writer_bpp = 3;    \
               code                         \
            }                               \
            break;                          \
         case 2:                            \
            {                               \
               const int reader_bpp = 2;    \
               const int writer_bpp = 2;    \
               code                         \
            }                               \
            break;                          \
         case 1:                            \
            {                               \
               const int reader_bpp = 1;    \
               const int writer_bpp = 1;    \
               code                         \
            }                               \
            break;                          \
         default:                           \
            break;                          \
      }
```
then, write a generic routines like:

 ```
   pixels_src = surf->pixels;
   pixels_dst = surf_ret->pixels;

   PIXEL_SWITCH_BPP(
      for (y = 0; y < h; y++) {
         for (x = 0; x < w; x++) {
            uint32_t color = PIXEL_READER(pixels_src, x);
            uint32_t dest_color;
            if (color == color_red) {
               dest_color = color_green;
            } else {
               dest_color = color;
            }
            PIXEL_WRITER(pixels_dst, x, dest_color);
         }
         pixels_src += surf->pitch;
         pixels_dst += surf_ret->pitch;
      }
   );
```

InfoTeddy commented:
I'm in favor of having these functions/macros be public, because VVVVVV currently rolls its own read/write pixel functions.

bitset32 commented:
Forgive my ignorance, but could the switch statement in SDL_ReadPixelValue:

```C
Uint32 SDL_ReadPixelValue(void *data, SDL_PixelFormat *format)
{
    Uint8 *src_px = (Uint8 *)data;

    /* Return the correct number of bytes in the correct endianness */
    switch(format->BytesPerPixel) {
        case 1:
            return *src_px;
        case 2:
            return *(Uint16*)src_px;
        case 3:
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
            return ((Uint32)src_px[0] << 16) | ((Uint32)src_px[1] << 8) | ((Uint32)src_px[2] << 0);
#else
            return ((Uint32)src_px[2] << 16) | ((Uint32)src_px[1] << 8) | ((Uint32)src_px[0] << 0);
#endif
        case 4:
            return *(Uint32*)src_px;
    }

    /* Invalid pixel format. */
    return 0;
}
```

Be replaced with a memcpy:

```C
Uint32 SDL_ReadPixelValue(void *data, SDL_PixelFormat *format)
{
    Uint32 pixel = 0;
    memcpy(&pixel, data, format->BytesPerPixel);
    return pixel;
}
```

InfoTeddy commented on include/SDL_pixels.h:
```diff
@@ -627,6 +627,57 @@ extern DECLSPEC void SDLCALL SDL_GetRGBA(Uint32 pixel,
+ * \param format an SDL_PixelFormat structure describing the format of the pixel
+ * \param value a pixel value in the format you'd get from SDL_MapRGB or SDL_MapRGBA
+ *
+ * \since This function is available since SDL 2.0.18
```
```suggestion
 * \since This function is available since SDL 2.0.20
```

sulix commented:
> Forgive my ignorance, but could the switch statement in SDL_ReadPixelValue:
> 
> …
> 
> Be replaced with a memcpy:
> 
> ```c
> Uint32 SDL_ReadPixelValue(void *data, SDL_PixelFormat *format)
> {
>     Uint32 pixel = 0;
>     memcpy(&pixel, data, format->BytesPerPixel);
>     return pixel;
> }
> ```

I'd guess this wouldn't work as-is, as on big-endian systems, the address of the pixel data in the Uint32 will be padded. Maybe something like:
```c
Uint32 pixel = 0;
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
memcpy(((Uint8*)&pixel) + (4 - format->BytesPerPixel), data, format->BytesPerPixel);
#else
memcpy(&pixel, data, format->BytesPerPixel);
#endif
return pixel;
```
could work, though…
